clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -157,6 +160,42 @@ public void testFetchNormal() {
     }
 
     @Test
+    public void testFetcherIgnoresControlRecords() {
+        subscriptions.assignFromUser(singleton(tp));
+        subscriptions.seek(tp, 0);
+
+        // normal fetch
+        assertEquals(1, fetcher.sendFetches());
+        assertFalse(fetcher.hasCompletedFetches());
+
+        ByteBuffer buffer = ByteBuffer.allocate(1024);
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, CompressionType.NONE, TimestampType.CREATE_TIME, 0L);
+        builder.append(0L, "key".getBytes(), null);
+        builder.appendControlRecord(0L, ControlRecordType.COMMIT, null);
+        builder.append(0L, "key".getBytes(), null);
+        builder.close();
+
+        builder = MemoryRecords.builder(buffer, CompressionType.NONE, TimestampType.CREATE_TIME, 3L);
+        builder.appendControlRecord(0L, ControlRecordType.ABORT, null);
+        builder.close();
+
+        buffer.flip();
+
+        client.prepareResponse(fetchResponse(MemoryRecords.readableRecords(buffer), Errors.NONE, 100L, 0));
+        consumerClient.poll(0);
+        assertTrue(fetcher.hasCompletedFetches());
+
+        Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords = fetcher.fetchedRecords();
+        assertTrue(partitionRecords.containsKey(tp));
+
+        List<ConsumerRecord<byte[], byte[]>> records = partitionRecords.get(tp);
+        assertEquals(2, records.size());
+        assertEquals(3L, subscriptions.position(tp).longValue());
```

Comment:
Should have known not to try to slip something by Jun! Fixing this is actually a bit tricky since we need to retain the control record long enough to update the consumer's position. This requires a bit of refactoring in `Fetcher`, which I had intended to do in a follow-up. I'll see if there's an easier workaround for now.